### PR TITLE
win: specifying the initial directory

### PR DESCRIPTION
**Pull request by secondarykey**
Thanks for the useful package!

I implemented it so that the initial directory can be specified.(DirectoryBuilder)

Added callback function to set when DirectoryBuilder.StartDir is specified.
"BFFM_*" constants should be in the w32 package,but it was difficult(fork repository) so I wrote it once.

I tried to implement DirectoryBuilder.SetStartDir(), but please delete it if it interferes with other OS.

(Please excuse my poor English.

### CODE REVIEW

**Review by sqweek — COMMENTED** [review]
Thanks for your work, this looks great! I've made a couple of minor comments regarding things I'd like changed before merging, let me know if you don't have time to address these and I'll try to get to them over the next week or so.

File: dlgs.go [hunk 1]
```diff
@@ -144,3 +144,9 @@ func (b *DirectoryBuilder) Title(title string) *DirectoryBuilder {
 	b.Dlg.Title = title
 	return b
 }
+
+// StartDir specifies the initialize directory to be used for the dialog.
```
[the remark this answers is not in the document]
Comment (secondarykey) [reply dlgs.go]: Thank you for your review.
I don't seem to be studying English at all.

File: dlgs_windows.go [hunk 2]
```diff
@@ -123,11 +123,40 @@ type dirdlg struct {
...
+	BFFM_SETSTATUSTEXTA  = (w32.WM_USER + 100)
+	BFFM_SETSTATUSTEXTW  = (w32.WM_USER + 104)
+	BFFM_ENABLEOK        = (w32.WM_USER + 101)
+	BFFM_SETSELECTIONA   = (w32.WM_USER + 102)
+	BFFM_SETSELECTIONW   = (w32.WM_USER + 103)
+	BFFM_SETOKTEXT       = (w32.WM_USER + 105)
+	BFFM_SETEXPANDED     = (w32.WM_USER + 106)
+	BFFM_SETSTATUSTEXT   = BFFM_SETSTATUSTEXTW
+	BFFM_SETSELECTION    = BFFM_SETSELECTIONW
+	BFFM_VALIDATEFAILED  = BFFM_VALIDATEFAILEDW
```
[the remark this answers is not in the document]
Comment (secondarykey) [reply dlgs_windows.go]: That's right!
I tried to make only the prefix lowercase, but what about?

File: dlgs_windows.go [hunk 3]
```diff
@@ -123,11 +123,40 @@ type dirdlg struct {
...
+)
+
+func callbackDefaultDir(hwnd w32.HWND, msg uint, lParam, lpData uintptr) int {
+	if msg == BFFM_INITIALIZED {
```
[the remark this answers is not in the document]
Comment (secondarykey) [reply dlgs_windows.go]: Sure, it seems a little easier to do, but as far as I've read the Microsoft documentation, it looks like a callback is needed.
https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/bb762598(v=vs.85)#remarks

**Comment by sqweek** [comment]
Thanks again @secondarykey! Sorry this took so long, work got real busy there for a few weeks.

File: dlgs_windows.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sqweek) [reply dlgs_windows.go]: Yes I agree that seems to be the only way to do it, my comment was trying to communicate my surprise that no simpler API was provided :)

**Comment by secondarykey** [comment]
Thank you for merging.
It was fast enough.There are times when Everyone is busy.

**Pull request closed without merging**